src/AstTypeAnalysis.cpp

```diff
@@ -476,251 +432,244 @@ AstClause* createAnnotatedClause(
+        if (auto expr = agg.getTargetExpression()) {
+            addConstraint(isSubtypeOf(getVar(expr), getVar(agg)));
+            addConstraint(isSubtypeOf(getVar(agg), getVar(expr)));
+        }
+    }
 
-            for (const AstArgument* value : init.getArguments()) {
-                addConstraint(isSubtypeOfComponent(getVar(value), rec, i++));
-            }
+    /**
+     * Utility function.
+     * Iterate over atoms valid pairs of (argument, type-attribute) and apply procedure `map` for its
+     * side-effects.
+     */
+    void iterateOverAtom(const AstAtom& atom, std::function<void(const AstArgument&, const Type&)> map) {
+        // get relation
+        auto rel = getAtomRelation(&atom, &program);
+        if (rel == nullptr) {
+            return;  // error in input program
         }
 
-        // visit aggregates
-        void visitAggregator(const AstAggregator& agg) override {
-            if (agg.getOperator() == AggregateOp::COUNT) {
-                addConstraint(isSubtypeOf(getVar(agg), env.getNumberType()));
-            } else if (agg.getOperator() == AggregateOp::MEAN) {
-                addConstraint(isSubtypeOf(getVar(agg), env.getFloatType()));
-            } else {
-                addConstraint(hasSuperTypeInSet(getVar(agg), env.getNumericTypes()));
-            }
+        auto atts = rel->getAttributes();
+        auto args = atom.getArguments();
+        if (atts.size() != args.size()) {
+            return;  // error in input program
+        }
 
-            // If there is a target expression - it should be of the same type as the aggregator.
-            if (auto expr = agg.getTargetExpression()) {
-                addConstraint(isSubtypeOf(getVar(expr), getVar(agg)));
-                addConstraint(isSubtypeOf(getVar(agg), getVar(expr)));
+        for (size_t i = 0; i < atts.size(); i++) {
+            const auto& typeName = atts[i]->getTypeName();
+            if (typeEnv.isType(typeName)) {
+                map(*args[i], typeEnv.getType(typeName));
             }
         }
-    };
+    }
+};
+
+std::map<const AstArgument*, TypeSet> TypeAnalysis::analyseTypes(const TypeEnvironment& typeEnv,
+        const AstClause& clause, const AstProgram& program, std::ostream* logs) {
+    return TypeConstraintsAnalysis(typeEnv, program).analyse(clause, logs);
+}
 
-    // run analysis
-    return Analysis(env, program).analyse(clause, logs);
+void TypeAnalysis::print(std::ostream& os) const {
+    os << "-- Analysis logs --" << std::endl;
+    os << analysisLogs.str() << std::endl;
+    os << "-- Result --" << std::endl;
+    for (const auto& cur : annotatedClauses) {
+        os << *cur << std::endl;
+    }
+}
+
+void TypeAnalysis::run(const AstTranslationUnit& translationUnit) {
+    // Check if debugging information is being generated
+    std::ostream* debugStream = nullptr;
+    if (Global::config().has("debug-report") || Global::config().get("show") == "type-analysis") {
```

Comment:
You can use `Global::config::has(key, value)` - something like:
```suggestion
    if (Global::config().has("debug-report") || Global::config().has("show", "type-analysis")) {
```